packages/cozy-client/src/models/konnectorFolder.js

```diff
@@ -244,3 +281,157 @@ export const buildFolderPermission = folder => {
+    )
+  )
+
+/**
+ * Test if a given folder has a given source account identifier reference
+ *
+ * @param {import('../types').IOCozyFolder} folder - folder to test
+ * @param {string} sourceAccountIdentifier - source account identifier
+ * @returns {boolean}
+ */
+const hasSourceAccountIdentifierReference = (
+  folder,
+  sourceAccountIdentifier
+) => {
+  return Boolean(
+    folder.referenced_by?.find(
+      ref =>
+        ref.type === 'io.cozy.accounts.sourceAccountIdentifier' &&
+        ref.id === sourceAccountIdentifier
+    )
+  )
+}
+
+/**
+ * Ensure that a given folder has the given konnector reference
+ *
+ * @param {object} options - options object
+ * @param {CozyClient} options.client - CozyClient instance
+ * @param {import('../types').IOCozyFolder} options.folder - folder to test
+ * @param {import('../types').IOCozyKonnector} options.konnector - konnector to which we want the reference
+ * @returns {Promise<void>}
+ */
+const ensureKonnectorReference = async ({ client, folder, konnector }) => {
+  const fileCollection = client.collection('io.cozy.files')
+  if (!hasKonnectorReference(folder, konnector.slug)) {
+    await fileCollection.addReferencesTo(konnector, [folder])
+  }
+}
+
+/**
+ * Ensure that a given folder has the given source account identifier reference
+ *
+ * @param {object} options - options object
+ * @param {CozyClient} options.client - CozyClient instance
+ * @param {import('../types').IOCozyFolder} options.folder - folder to test
+ * @param {string} options.sourceAccountIdentifier - source account identifier to which we want the reference
+ * @returns {Promise<void>}
+ */
+const ensureSourceAccountIdentifierReference = async ({
+  client,
+  folder,
+  sourceAccountIdentifier
+}) => {
+  const fileCollection = client.collection('io.cozy.files')
+  if (!hasSourceAccountIdentifierReference(folder, sourceAccountIdentifier)) {
+    await fileCollection.addReferencesTo(
+      {
+        _id: sourceAccountIdentifier,
+        _type: 'io.cozy.accounts.sourceAccountIdentifier'
+      },
+      [folder]
+    )
+  }
+}
+
+/**
+ * try to find a konnector account folder using file references
+ *
+ * @param {object} options - options object
+ * @param {CozyClient} options.client - CozyClient instance
+ * @param {string} options.slug - konnector slug
+ * @param {string} options.sourceAccountIdentifier - source account identifier
+ * @returns {Promise<import('../types').IOCozyFolder>}
+ */
+export const findKonnectorAccountFolderByReference = async ({
+  client,
+  slug,
+  sourceAccountIdentifier
+}) => {
+  const fileCollection = client.collection('io.cozy.files')
+  const resp = await fileCollection.findReferencedBy({
+    _type: 'io.cozy.konnectors',
+    _id: `io.cozy.konnectors/${slug}`
+  })
+  const files = resp.included
+  return files.find(
+    file =>
+      file.type === 'directory' &&
+      !isTrashedFolder(file) &&
+      hasSourceAccountIdentifierReference(file, sourceAccountIdentifier)
+  )
+}
```

Comment:
You are bypassing all the cozy-client built-in query features, which is not good :wink: 
First, why using collection and not cozy-client query ? e.g. `client.query(Q('io.cozy.files').referencedBy({...})`
This way, you will benefit from the store

Second, you query the references, and after that filter the retrieved values, while it is precisely the client.query job.
So, that would make something like this: 

```js
const resp = await client.query(Q('io.cozy.files').partialIndex({type: "directory", trashed: false}.referencedBy({...})
```

For the `hasSourceAccountIdentifierReference`, I'm not sure I understand correctly, the file must have a `io.cozy.konnectors` reference + `io.cozy.accounts.sourceAccountIdentifier`? And I'm surprised about this last one, is it a new doctype?

Reply:
I did not know about this `Q().referencedBy`, I just took the same method used to query magic folders in [folder.js](https://github.com/cozy/cozy-client/blob/0bc07eb5b637c00bd5e7e96e6acf68b57ad02b5c/packages/cozy-client/src/models/folder.js#L93). Ok, let's use cozy-client query (And I will update folder.js.

And second, I'm surprised the request you propose works, since the collection findReferencedBy uses a specific stack request but I see there are helpers for this in the QueryDefinition. let's do this then, there will be a lot less code :+1:.